File: angular-client/akira-ng/projects/president/src/app/game/deck.ts
import { Card, RANKS, SUITS } from './card';
import { Rng } from './rng';

/** Builds `numDecks` standard 52-card decks (no jokers) with unique uids. */
export function buildDeck(numDecks = 2): Card[] {
  const cards: Card[] = [];
  for (let copy = 0; copy < numDecks; copy++) {
    for (const suit of SUITS) {
      for (const rank of RANKS) {
        cards.push({ rank, suit, uid: `${rank}${suit}#${copy}` });
      }
    }
  }
  return cards;
}

/**
 * Shuffles a fresh two-deck shoe and deals it round-robin to `playerCount`
 * players. Hands may differ in size by one card when 104 doesn't divide evenly.
 */
export function deal(playerCount: number, rng: Rng, numDecks = 2): Card[][] {
  const shoe = rng.shuffle(buildDeck(numDecks));
  const hands: Card[][] = Array.from({ length: playerCount }, () => []);
  shoe.forEach((card, i) => hands[i % playerCount].push(card));
  return hands;
}

File: spring-decks/src/main/java/com/gpt/decks/engine/president/Deck.java
package com.gpt.decks.engine.president;

import java.util.ArrayList;
import java.util.List;

/** Builds N standard decks (no jokers) and deals them round-robin. */
public final class Deck {

  private Deck() {
  }

  public static List<Card> build(int numDecks) {
    List<Card> cards = new ArrayList<>();
    for (int copy = 0; copy < numDecks; copy++) {
      for (Suit suit : Suit.values()) {
        for (Rank rank : Rank.values()) {
          cards.add(new Card(rank, suit, rank.label() + suit.label() + "#" + copy));
        }
      }
    }
    return cards;
  }

  public static List<List<Card>> deal(int players, Rng rng, int numDecks) {
    List<Card> shoe = build(numDecks);
    rng.shuffle(shoe);
    List<List<Card>> hands = new ArrayList<>();
    for (int i = 0; i < players; i++) {
      hands.add(new ArrayList<>());
    }
    for (int i = 0; i < shoe.size(); i++) {
      hands.get(i % players).add(shoe.get(i));
    }
    return hands;
  }
}
